Basic Finance Math

dean-krueger opened the pull request:
# Summary of Changes

This PR adds a set of basic financial math functions to the Economic Entity class, which will hopefully be useful in covering most finical math operations for all future functions, though they may need to be expanded in the future.

# Related CEPs and Issues

This PR is related to:

- Closes #1863 

# Associated Developers

@gonuke

# Design Notes

@gonuke laid out the set of functions to be added in Issue #1863, and this is hopefully a fairly straightforward implementation of those functions.

# Testing and Validation

Tests were added to agent_tests for each of these new functions where they are validated by hand calcs. All tests passed after rebuilding cyclus, and I manually verified that all four new tests were called in that set of unit tests to be sure I did it right.

# Checklist

 - [x] Read the [Contributing to Cyclus](https://fuelcycle.org/kernel/contributing_to_cyclus.html) guide.
 - [x]  Compile and run locally.
 - [x]  Add or update tests.
 - [x]  Document if needed.
 - [x]  Follow style guidelines.
 - [x]  Update the changelog.
 - [x]  Address all review comments.
Reviewers, please refer to the Cyclus [Guide for Reviewers](https://fuelcycle.org/kernel/pr_review.html).

dean-krueger committed:
added core and packages.xml to gitignore

dean-krueger committed:
Added basic finance math functions and appropriate tests

dean-krueger committed:
updated CHANGELOG

dean-krueger commented:
<img width="450" alt="Screenshot 2025-06-14 at 11 36 38 AM" src="https://github.com/user-attachments/assets/07eadcb7-c86c-463e-808e-9d937cfbc434" />

Here is a nicely formatted version of the functions I implemented, to hopefully make the review process a little easier. These are the forms I used to do my hand calculations in the tests.

gonuke commented on tests/agent_tests/agent_tests.cc:
```diff
@@ -78,3 +78,27 @@ TEST_P(AgentTests, SetThenGetEconParam) {
 }
 
+TEST_P(AgentTests, PV) {
+  double present_value = agent_->PV(10, 0.05, 10000.0, 100.0);
```
Probably makes sense to test with only F, only A, with both, and with i = 0, i != 0, n<0, n=0, n>0

gonuke commented on tests/agent_tests/agent_tests.cc:
```diff
@@ -78,3 +78,27 @@ TEST_P(AgentTests, SetThenGetEconParam) {
+  EXPECT_NEAR(present_value, pv_hand_calc, 1e-3);
+}
+
+TEST_P(AgentTests, FV) {
```
ibid,regarding variety of testing

gonuke commented on tests/agent_tests/agent_tests.cc:
```diff
@@ -78,3 +78,27 @@ TEST_P(AgentTests, SetThenGetEconParam) {
+}
+
+TEST_P(AgentTests, PMT) {
+  double payment = agent_->PMT(10, 0.05, 10000.0, 40000.0);
```
This can be tested against PV and FV

PV(n, i, 0, PMT(n, i, 1, 0)) should equal 1

gonuke commented on tests/agent_tests/agent_tests.cc:
```diff
@@ -78,3 +78,27 @@ TEST_P(AgentTests, SetThenGetEconParam) {
+}
+
+TEST_P(AgentTests, PV_def_by_A) {
+  std::vector<double> payments = {100.0, 200.0, 300.0, 10.0, 15.0, 1000.0};
```
Can also be tested against prior functions

gonuke commented on src/economic_entity.h:
```diff
@@ -32,6 +32,97 @@ class EconomicEntity {
+virtual double PV(double i, const std::vector<double>& A) const {
+    double pv = 0.0;
+
+    for (int t = 0; t < A.size(); ++t) {
+        pv += A[t] / std::pow((1 + i), t);
```
There is almost certainly an off by one error of some kind here 

the contribution from the first payment `A[0]` should happen at the end of the first time period such that the PV factor is at `t=1`

You'd probably catch this if testing this function with a set of constant payments and comparing to PV

dean-krueger replied:
Ahh, that was a slight misunderstanding on my part, which is reflected in the way I was testing (my hand calc has the first exponent as a 0 not a 1). I'll fix this and then add those extra tests you were talking about to hopefully catch anything else like this.

gonuke reviewed (changes requested):
Thanks @dean-krueger.  I have a few comments on testing and one question about the implementation of the varying payments

gonuke commented:
Regarding the formatted equations, the last one should go from t=[1,n] by convention

dean-krueger committed:
added a few tests to more thoroughly cover some scenarios

dean-krueger commented:
Not sure if the conda/Mac test is still failing. It did last time, but I've gone and added a bunch of those tests you asked for (good idea by the way, more tests is almost always better), and fixed the off-by-one error. Let me know if there are other things you see in your review, and I'll work on fixing them ASAP. If that one workflow test is still failing I can investigate why later.

gonuke commented on tests/agent_tests/agent_tests.cc:
```diff
@@ -78,3 +79,93 @@ TEST_P(AgentTests, SetThenGetEconParam) {
+  EXPECT_NEAR(present_value, pv_hand_calc, 1e-3);
+}
+
+TEST_P(AgentTests, PV_only_F) {
```
You can probably combine a few different tests into a single function:
* PV_only_F
* PV_only_A
* PV_A_and_F

Could also try a few different values of i and n.  So `test_PV` could be:
* i=0, n=0, F=1, A=0
* i>0, n=1, F=1, A=0
* i>0, n>1, F=1, A=0
* i>0, n>1, F>1, A=0
* i>0, n>1, F=0, A=1
* i>0, n>1, F=0, A>1
* i>0, n>1, F>1, A>1
Etc

It's also valid to test them against a computed response rather than a hand calculated response. E.g.
`pv(2,0.01,1,0) == pow(1.01,2)`

gonuke reviewed (changes requested):
I think it probably makes sense to discuss the desired behavior and broad testing strategy in person

dean-krueger commented:
Sounds good. I'll stick a pin in this until Wednesday and make it an item for the Fuel Cycles meeting so we can talk about it.

gonuke committed:
propose finance math tests

gonuke committed:
add test for varying A

gonuke committed:
make second series not constant

gonuke committed:
add def'n of near

gonuke committed:
fix typos

dean-krueger committed:
Merge pull request #1 from gonuke/finance-math

Consolidated testing for basic finance math

dean-krueger committed:
removed throw on n < 0 condition, still failing some tests, but now they're actually failing and not just throwing

gonuke commented on tests/agent_tests/agent_tests.cc:
```diff
@@ -78,3 +79,109 @@ TEST_P(AgentTests, SetThenGetEconParam) {
+  EXPECT_NEAR(F / std::pow((1 + (i/12)), n*12), agent_->PV(n*12,i/12,F,0), finance_eps); // general
+  EXPECT_NEAR(F / std::pow((1+i), -n), agent_->PV(-n,i,F,0), finance_eps); // negative time
+
+  EXPECT_NEAR(A, agent_->PV(n,0,0,A), finance_eps); // discount rate = 0
```
My bad:
```suggestion
  EXPECT_NEAR(A * n, agent_->PV(n,0,0,A), finance_eps); // discount rate = 0
```

gonuke commented on tests/agent_tests/agent_tests.cc:
```diff
@@ -78,3 +79,109 @@ TEST_P(AgentTests, SetThenGetEconParam) {
+  EXPECT_NEAR(A * (1 - std::pow((1 + (i/12)), -n*12)) / (i/12), agent_->PV(n*12,i/12,0,A), finance_eps); // general
+  EXPECT_NEAR(A * (1 - std::pow((1 + i), n)) / i, agent_->PV(-n,i,0,A), finance_eps); // negative time
+
+  EXPECT_NEAR(F + A, agent_->PV(n,0,F,A), finance_eps); // discount rate = 0
```
Same error as above
```suggestion
  EXPECT_NEAR(F + A * n, agent_->PV(n,0,F,A), finance_eps); // discount rate = 0
```

gonuke commented on tests/agent_tests/agent_tests.cc:
```diff
@@ -78,3 +79,109 @@ TEST_P(AgentTests, SetThenGetEconParam) {
+  // testing by "round trip" with already tested functions
+  EXPECT_NEAR(P, agent_->PV(n,0,agent_->FV(n,0,P,0),0), finance_eps);  // discount rate = 0
+  EXPECT_NEAR(P, agent_->PV(0,i,agent_->FV(0,i,P,0),0), finance_eps);  // discount time = 0
+  EXPECT_NEAR(P, agent_->PV(i,i,agent_->FV(1,i,P,0),0), finance_eps);  // discount time = 1
```
type
```suggestion
  EXPECT_NEAR(P, agent_->PV(1,i,agent_->FV(1,i,P,0),0), finance_eps);  // discount time = 1
```

gonuke commented on tests/agent_tests/agent_tests.cc:
```diff
@@ -78,3 +79,109 @@ TEST_P(AgentTests, SetThenGetEconParam) {
+  EXPECT_NEAR(P, agent_->PV(-n,i,agent_->FV(-n,i,P,0),0), finance_eps);  // general
+  
+  // manual testing
+  EXPECT_NEAR(A, agent_->FV(n,0,0,A), finance_eps); // discount rate = 0
```
repeated error
```suggestion
  EXPECT_NEAR(A * n, agent_->FV(n,0,0,A), finance_eps); // discount rate = 0
```

gonuke commented on tests/agent_tests/agent_tests.cc:
```diff
@@ -78,3 +79,109 @@ TEST_P(AgentTests, SetThenGetEconParam) {
+  EXPECT_NEAR(A, agent_->FV(n,0,0,A), finance_eps); // discount rate = 0
+  EXPECT_NEAR(0, agent_->FV(0,i,0,A), finance_eps);  // discount time = 0
+  EXPECT_NEAR(A, agent_->FV(1,i,0,A), finance_eps); // discount time = 1
+  EXPECT_NEAR(A * (std::pow((1 + i), n) - 1) / i, agent_->PV(n,i,0,A), finance_eps); // general
+  EXPECT_NEAR(A * (std::pow((1 + i/12), n*12) - 1) / i/12, agent_->PV(n*12,i/12,0,A), finance_eps); // general
+  EXPECT_NEAR(A * (std::pow((1 + i), -n) - 1) / i, agent_->PV(-n,i,0,A), finance_eps); // general
```
Dumb copy/paste errors on my part
```suggestion
  EXPECT_NEAR(A * (std::pow((1 + i), n) - 1) / i, agent_->FV(n,i,0,A), finance_eps); // general
  EXPECT_NEAR(A * (std::pow((1 + i/12), n*12) - 1) / i/12, agent_->FV(n*12,i/12,0,A), finance_eps); // general
  EXPECT_NEAR(A * (std::pow((1 + i), -n) - 1) / i, agent_->FV(-n,i,0,A), finance_eps); // general
```

gonuke commented on tests/agent_tests/agent_tests.cc:
```diff
@@ -78,3 +79,109 @@ TEST_P(AgentTests, SetThenGetEconParam) {
+  EXPECT_NEAR(A * (std::pow((1 + i/12), n*12) - 1) / i/12, agent_->PV(n*12,i/12,0,A), finance_eps); // general
+  EXPECT_NEAR(A * (std::pow((1 + i), -n) - 1) / i, agent_->PV(-n,i,0,A), finance_eps); // general
+
+  EXPECT_NEAR(P + A, agent_->FV(n,0,P,A), finance_eps); // discount rate = 0
```
repeated error
```suggestion
  EXPECT_NEAR(P + A * n, agent_->FV(n,0,P,A), finance_eps); // discount rate = 0
```

gonuke commented on tests/agent_tests/agent_tests.cc:
```diff
@@ -78,3 +79,109 @@ TEST_P(AgentTests, SetThenGetEconParam) {
+  EXPECT_NEAR(P + A, agent_->FV(n,0,P,A), finance_eps); // discount rate = 0
+  EXPECT_NEAR(P, agent_->FV(0,i,P,A), finance_eps); // discount time = 0
+  EXPECT_NEAR(P * (1+i) + A, agent_->FV(1,i,P,A), finance_eps); // discount time = 1
+  EXPECT_NEAR(P * std::pow((1+i), n) + A * (std::pow((1 + i), n) - 1),
```
```suggestion
  EXPECT_NEAR(P * std::pow((1+i), n) + A * (std::pow((1 + i), n) - 1) / i,
```

gonuke commented on tests/agent_tests/agent_tests.cc:
```diff
@@ -78,3 +79,109 @@ TEST_P(AgentTests, SetThenGetEconParam) {
+  // all testing by "round trip" with already tested functions
+
+  EXPECT_NEAR(P, agent_->PV(n,0,0,agent_->PMT(n,0,P,0)), finance_eps); // discount rate = 0
+  EXPECT_NEAR(P, agent_->PV(0,i,0,agent_->PMT(0,i,P,0)), finance_eps); // discount time = 0
```
The left hand side evaluates to `nan` - this case may not be well defined and we should discuss the right solution.

Basically, there is no correct answer for how big the equivalent payment is when there are 0 payment.  So maybe this is a case where we want to make sure n >= 0 and otherwise throw an exception.

gonuke commented on tests/agent_tests/agent_tests.cc:
```diff
@@ -78,3 +79,109 @@ TEST_P(AgentTests, SetThenGetEconParam) {
+  EXPECT_NEAR(P, agent_->PV(n,i,0,agent_->PMT(n,i,P,0)), finance_eps); // general
+  
+  EXPECT_NEAR(F, agent_->FV(n,0,0,agent_->PMT(n,0,0,F)), finance_eps); // discount rate = 0
+  EXPECT_NEAR(F, agent_->FV(0,i,0,agent_->PMT(0,i,0,F)), finance_eps); // discount time = 0
```
same error as above with `nan`

gonuke commented on tests/agent_tests/agent_tests.cc:
```diff
@@ -78,3 +79,109 @@ TEST_P(AgentTests, SetThenGetEconParam) {
+  EXPECT_NEAR(agent_->PV(n,i,0,A1[0]), agent_->PV(i, A1), finance_eps);
+  double PV_A2 = 0;
+  for (int n2=0; n2 < A2.size(); n2++) {
+    PV_A2 += agent_->PV(n2+1, i, 0, A2[n2]);
```
Each payment acts like a separate single future value in this case
```suggestion
    PV_A2 += agent_->PV(n2+1, i, A2[n2], 0);
```

gonuke reviewed (changes requested):
Now that I've actually seen the tests run (my bad for not doing that before), I've found all of my own errors here.

dean-krueger commented:
Well that's a convenient development! I was just about to go dig through these myself, but thanks for taking another look at them. I'll make these changes and then see where we stand with the tests.

dean-krueger committed:
Update tests/agent_tests/agent_tests.cc

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>
Signed-off-by: Dean Krueger <74223179+dean-krueger@users.noreply.github.com>

dean-krueger committed:
Update tests/agent_tests/agent_tests.cc

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>
Signed-off-by: Dean Krueger <74223179+dean-krueger@users.noreply.github.com>

dean-krueger committed:
Update tests/agent_tests/agent_tests.cc

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>
Signed-off-by: Dean Krueger <74223179+dean-krueger@users.noreply.github.com>

dean-krueger committed:
Update tests/agent_tests/agent_tests.cc

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>
Signed-off-by: Dean Krueger <74223179+dean-krueger@users.noreply.github.com>

dean-krueger committed:
Update tests/agent_tests/agent_tests.cc

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>
Signed-off-by: Dean Krueger <74223179+dean-krueger@users.noreply.github.com>

dean-krueger committed:
Update tests/agent_tests/agent_tests.cc

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>
Signed-off-by: Dean Krueger <74223179+dean-krueger@users.noreply.github.com>

dean-krueger committed:
Update tests/agent_tests/agent_tests.cc

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>
Signed-off-by: Dean Krueger <74223179+dean-krueger@users.noreply.github.com>

dean-krueger committed:
Update tests/agent_tests/agent_tests.cc

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>
Signed-off-by: Dean Krueger <74223179+dean-krueger@users.noreply.github.com>

dean-krueger committed:
Fixed tests, added catch to PMT for when n <= 0 --> infinity or undefinted payment

gonuke commented on src/economic_entity.h:
```diff
@@ -32,6 +32,106 @@ class EconomicEntity {
+    if (i != 0) {
+        return (p_term + f_term) * i / (1 - std::pow((1 + i), -n));
+    } else {
+        return (p_term + F) / n;
```
Unless you reject this suggestion, I'll merge when this has been committed

dean-krueger committed:
Update src/economic_entity.h

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>
Signed-off-by: Dean Krueger <74223179+dean-krueger@users.noreply.github.com>

gonuke reviewed (approved):
Thanks @dean-krueger

dean-krueger committed:
Merge branch 'main' into finance-math

Signed-off-by: Dean Krueger <74223179+dean-krueger@users.noreply.github.com>
